gog/graph.go

```diff
@@ -99,111 +92,103 @@ func (g *Grapher) Graph(files []*ast.File, typesPkg *types.Package, typesInfo *t
+	// Transfer pkg graph data to output
+	g.Defs = append(g.Defs, v.pkgDefs...)
+	g.Refs = append(g.Refs, v.pkgRefs...)
+	g.Docs = append(g.Docs, v.pkgDocs...)
 
-		if !isPkg {
-			def, err := g.NewDef(obj, ident)
-			if err != nil {
-				return err
-			}
-			pkgDefs = append(pkgDefs, def)
-		}
+	return nil
+}
 
-		ref := g.NewRef(ident, obj, typesPkg.Path())
-		ref.IsDef = true
-		pkgRefs = append(pkgRefs, ref)
-	}
+type astVisitor struct {
+	g         *Grapher
+	typesPkg  *types.Package
+	typesInfo *types.Info
 
-	for ident, obj := range typesInfo.Uses {
-		if _, isLabel := obj.(*types.Label); isLabel {
-			g.skipResolve[ident] = struct{}{}
-			continue
-		}
+	err     error
+	pkgDefs []*Def
+	pkgRefs []*Ref
+	pkgDocs []*Doc
+}
 
-		if obj == nil || ident == nil || ident.Name == "_" {
-			continue
-		}
+func (v *astVisitor) Visit(node ast.Node) (w ast.Visitor) {
+	if v.err != nil {
+		return nil
+	}
 
-		if _, skip := skipResolveObjs[obj]; skip {
-			g.skipResolve[ident] = struct{}{}
+	switch n := node.(type) {
+	case *ast.File:
+		// Create a ref that represent the name of the package ("package foo")
+		pkgObj := types.NewPkgName(n.Name.Pos(), v.typesPkg, v.typesPkg.Name(), v.typesPkg)
+		ref := v.g.NewRef(n.Name, pkgObj, v.typesPkg.Path())
+		v.pkgRefs = append(v.pkgRefs, ref)
+
+	case *ast.ImportSpec:
+		if obj := v.typesInfo.Implicits[n]; obj != nil {
+			ref := v.g.NewRef(n, obj, v.typesPkg.Path())
+			v.pkgRefs = append(v.pkgRefs, ref)
 		}
 
-		if _, seen := seen[ident]; seen {
-			continue
+	case *ast.Ident:
+		ident := n
+		if ident.Name == "_" {
+			break
 		}
 
-		if _, isLabel := obj.(*types.Label); isLabel {
-			continue
+		isDef := false
+		if obj := v.typesInfo.Defs[ident]; obj != nil {
+			isDef = true
+			// don't treat import aliases as things that belong to this package
+			if _, isPkg := obj.(*types.PkgName); !isPkg {
+				def, err := v.g.NewDef(obj, ident)
+				if err != nil {
+					v.err = err
+					return nil
+				}
+				v.pkgDefs = append(v.pkgDefs, def)
+			}
 		}
 
-		ref := g.NewRef(ident, obj, typesPkg.Path())
-		pkgRefs = append(pkgRefs, ref)
-	}
+		if obj := v.typesInfo.ObjectOf(ident); obj != nil {
+			ref := v.g.NewRef(ident, obj, v.typesPkg.Path())
+			ref.IsDef = isDef
+			v.pkgRefs = append(v.pkgRefs, ref)
+		}
 
-	// Create a ref that represent the name of the package ("package foo")
-	// for each file.
-	for _, f := range files {
-		pkgObj := types.NewPkgName(f.Name.Pos(), typesPkg, typesPkg.Name(), typesPkg)
-		ref := g.NewRef(f.Name, pkgObj, typesPkg.Path())
-		pkgRefs = append(pkgRefs, ref)
-	}
+	case *ast.LabeledStmt:
+		ast.Walk(v, n.Stmt)
+		return nil
 
-	if !g.SkipDocs {
-		pkgDocs, err = g.emitDocs(files, typesPkg, typesInfo)
-		if err != nil {
-			return err
-		}
+	case *ast.BranchStmt:
+		return nil
 	}
 
-	// Transfer pkg graph data to output
-	g.Defs = append(g.Defs, pkgDefs...)
-	g.Refs = append(g.Refs, pkgRefs...)
-	g.Docs = append(g.Docs, pkgDocs...)
-
-	return nil
+	return v
```

Comment:
I tried to review `refactored Grapher.Graph` closely, but not 100% convinced I did a good job. But I assume the test cases we have cover this very closely, so 👍 .
Also this is much more readable and understandable. nice change :)